pkg/controller/direct/firestore/firestoredocument_controller.go

```diff
@@ -0,0 +1,278 @@
+	mapCtx := &direct.MapContext{}
+	status.ObservedState = FirestoreDocumentObservedState_v1alpha1_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firestoreDocumentAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	fqn := a.id.String()
+
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating FirestoreDocument", "name", fqn)
+
+	resource := direct.ProtoClone(a.desired)
+
+	latest := a.actual
+
+	changedFieldPaths := changedFieldPaths(a.actual, a.desired)
+	if len(changedFieldPaths) > 0 {
+		resource.Name = fqn
+		req := &pb.UpdateDocumentRequest{
+			Document: resource,
+		}
+		updated, err := a.firestoreClient.UpdateDocument(ctx, req)
+		if err != nil {
+			return fmt.Errorf("updating FirestoreDocument %s: %w", fqn, err)
+		}
+		log.V(2).Info("successfully updated FirestoreDocument", "name", fqn)
+		latest = updated
+	}
+
+	status := &krm.FirestoreDocumentStatus{}
+	status.ExternalRef = direct.PtrTo(a.id.String())
+	mapCtx := &direct.MapContext{}
+	status.ObservedState = FirestoreDocumentObservedState_v1alpha1_FromProto(mapCtx, latest)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firestoreDocumentAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	fqn := a.id.String()
+	if a.actual == nil {
+		return nil, fmt.Errorf("FirestoreDocument %q not found", fqn)
+	}
+
+	mapCtx := &direct.MapContext{}
+	objSpec := FirestoreDocumentSpec_v1alpha1_FromProto(mapCtx, a.actual)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	obj := &krm.FirestoreDocument{
+		Spec: *objSpec,
+	}
+
+	obj.SetGroupVersionKind(krm.FirestoreDocumentGVK)
+	obj.Name = a.id.Document
+	obj.Spec.DatabaseRef.External = a.id.Parent.String()
+	if a.id.Collection != "" {
+		obj.Spec.Collection = direct.PtrTo(a.id.Collection)
+	}
+
+	unstructuredObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, fmt.Errorf("converting FirestoreDocument to unstructured failed: %w", err)
+	}
+
+	u := &unstructured.Unstructured{
+		Object: unstructuredObj,
+	}
+
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *firestoreDocumentAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	fqn := a.id.String()
+
+	log := klog.FromContext(ctx)
+	log.V(2).Info("deleting FirestoreDocument", "name", fqn)
+
+	req := &pb.DeleteDocumentRequest{
+		Name: fqn,
+	}
+	if err := a.firestoreClient.DeleteDocument(ctx, req); err != nil {
+		if direct.IsNotFound(err) {
+			// Return success if not found (assume it was already deleted).
+			log.V(2).Info("skipping delete for non-existent FirestoreDocument, assuming it was already deleted", "name", fqn)
+			return true, nil
+		}
+		return false, fmt.Errorf("deleting FirestoreDocument %s: %w", fqn, err)
```

Comment:
Nit: is there an edge case around getting an error calling delete on a document which is being deleted but is not yet gone?